### ISSUE TITLE
[Feature Request] Specify activity names at runtime

### ISSUE THREAD

**Issue by cgillum** [temporalio/sdk-dotnet#24]
### Is your feature request related to a problem? Please describe.

I have a need to create activities with names that aren't known until runtime. The current design requires me to use `[Activity(string)]` on a method or delegate, which only allows me to use activity names known at compile time.

As far as I can tell, there's no simple way for me to add an `ActivityAttribute` to a `Delegate` at runtime in .NET, short of doing some kind of dynamic type generation (which seems like overkill).

### Describe the solution you'd like

A simple solution would be to add a `TemporalWorkerOptions.AddActivity(string name, Delegate del)` overload that allows me to specify the activity name explicitly.

**Comment by cretz** [comment]
:+1: Yes we do need this (we also need "dynamic activities" that catch-all activity invocations that aren't explicitly registered, but that's unrelated).

We have a couple of options.

> A simple solution would be to add a `TemporalWorkerOptions.AddActivity(string name, Delegate del)` overload that allows me to specify the activity name explicitly.

This is one option, but there are two problems here: 1) I still want people to be able to use `List<Delegate>` when creating a worker for the common case, and 2) this is not future proofed against other activity options that may come about in the future.

I'm thinking about keeping `public IList<Delegate> Activities { get; set; } = new List<Delegate>();`, moving `Temporalio.Activity.ActivityAttribute.Definition` to public `Temporalio.Activity.ActivityDefinition` and having a new `public IList<ActivityDefinition> ActivityDefinitions { get; set; } = new List<ActivityDefinition>();` that I validate at worker instantiation time.

Thoughts?

**Comment by cgillum** [comment]
For my use-case, the ability to register dynamic activities as a catch-all would also solve my problem (in fact, that would make my current proof-of-concept simpler). Your suggestion to expose `ActivityDefinitions` publicly is also fine from my POV.

**Comment by cretz** [comment]
:+1: I have opened https://github.com/temporalio/sdk-dotnet/issues/27 (as part of general feature https://github.com/temporalio/features/issues/201). I will probably also expose the activity definition because it's unreasonable to require attributes be the only way to name activities.

**Comment by cretz** [comment]
@cgillum - Also consider for your use case, if reasonable, to have a single activity with the discriminator in the input instead of the name itself. Granted if you're building a general purpose library for others to use, I can understand why name is better (and it is clearer from a UI POV).

**Comment by cgillum** [comment]
Yeah, this is effectively a general-purpose library for others to use, so the single activity approach unfortunately won't work for me.

Thanks for opening the other issue - I'll keep an eye on that.

**Comment by cgillum** [comment]
@cretz do you have an idea on when you might have time to start work on this?

If not, do you accept external PR contributions?

**Comment by cretz** [comment]
@cgillum - I have already started on your first suggestion. I am right in the middle of workflow development on https://github.com/cretz/temporal-sdk-dotnet/tree/workflow-more-impl but I have exposed activity definition publicly already. But that code is a mess because it is mid development. Give me a little while longer and another alpha will be released with full workflow support (due to vacation and such, this may not be at least until the week of the 20th).

Also in the meantime, would love feedback on SDK approach/code in general. Though not quite yet on that branch :-) May want to wait until the next workflow PR is up hopefully this week (will be merging #29 before I open the next one since I'm daisy-chaining for feedback reasons).

As for full impl of #27 and dynamic activity support, I need to wait until after I finish this development on workflows.
